File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.py
class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        nums.sort()
        self.heap = []
        self.k = k
        while nums and len(self.heap) < k:
            heapq.heappush(self.heap, nums.pop())


    def add(self, val: int) -> int:
        heapq.heappush(self.heap, val)
        if len(self.heap) > self.k:
            heapq.heappop(self.heap)
        res = heapq.heappop(self.heap)
        heapq.heappush(self.heap, res)
        return res




        


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.cpp
static const bool a() {
    std::ios_base::sync_with_stdio(false);
    std::cout.tie(nullptr);
    std::cin.tie(nullptr);
    return true;
}
class KthLargest {
public:
    KthLargest(int k, vector<int>& nums) : k(k) {
        nums_.insert(nums_.end(), nums.begin(), nums.end());
        make_heap(nums_.begin(), nums_.end(), greater<>());
        while (nums_.size() > k) {
            pop_heap(nums_.begin(), nums_.end(), greater<>());
            nums_.pop_back();
        }
    }
    
    int add(int val) {
        nums_.push_back(val);
        push_heap(nums_.begin(), nums_.end(), greater<>());
        if (nums_.size() > k) {
            pop_heap(nums_.begin(), nums_.end(), greater<>());
            nums_.pop_back();
        }
        return nums_.front();
    }
    
private:
    vector<int> nums_;
    int k;
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */